### PR TITLE
Validation that port used for receiving autentication token from brow…

### LINKED ISSUE #176: [Bug]: Not possible to finish browser authentication process
### Please confirm that you have:

- [X] Searched [existing issues](.) to see if your issue is a duplicate. (If you’ve found a duplicate issue, feel free to add additional information in a comment on it.)
- [X] Reproduced the issue in [the latest CLI version](https://www.npmjs.com/package/@shopify/cli).

### In which of these areas are you experiencing a problem?

Other

### Expected behavior

After you have authenticated in the browser the CLI process should resume the execution of the command

### Actual behavior

After finishing the authentication process in the browser the CLI get stuck showing the message `Auto-open timed out. Open the login page: Log in to Shopify Partners...` 

### Stack trace

_No response_

### Reproduction steps

1. run dev, and need to authenticate (i.e. first time for an app, or using --reset)
2. push to open the browser but don't authenticate
3. finish the CLI process
4. finish authenticating -- won’t work because the CLI has stopped listening
5. try from step 1 again
6. once the browser is opened try to authenticate. The CLI never receives the response from the browser and always shows the message `Auto-open timed out. Open the login page: Log in to Shopify Partners...` 
7. Kill the process that listen to the port `3456` or restarting the machine solve the problem

### Operating System

Mac OS

### Shopify CLI version (check your project's `package.json` if you're not sure)

3.2.0

### Shell

_No response_

### Node version (run `node -v` if you're not sure)

_No response_

### What language and version are you using in your application?

_No response_

### PR DESCRIPTION
**Pull request by alvaro-shopify**
### WHY are these changes introduced?

Fixes #176 

Along the authorization process the CLI opens an http server to listen to the browser response with the identity token. If for any reason the fixed port used for the server '3456' is already in use, the CLI gets stuck waiting for the response so the process never finishes.

### WHAT is this pull request doing?
- Adds a validation before opening the browser to get the identity token. In case the port is in use the CLI, it prompts the user to terminate the process that is blocking it. In case, the user confirm the action, the process is killed and the authorization goes on opening the browser. On the other hand, if the user rejects finishing the process, the CLI closes gracefully without any error message

cancel_process_prompt

- Added dependency [kill-port-process](https://github.com/hilleer/kill-port-process)
- Added dependency [find-process](https://github.com/yibn2008/find-process)

### How to test your changes?

1. Logout to ensure the needing of executing the authorization process `yarn shopify auth logout`
2. Run the dev command `yarn shopify app dev`
3. Suspend the process (CTRL + Z) when the CLI is waiting for the browser response
4. Execute again the dev command `yarn shopify app dev`

In MacOS to kill the process that listen to a specific port use follow these steps:
1. `lsof -i -P | grep LISTEN | grep 3456`
2. `kill -9 <pid_from_lsof>`

### CODE REVIEW

**Comment by alvaro-shopify** [comment]
> Since the port `3456` is a port that's [hardcoded](https://github.com/Shopify/cli/blob/main/packages/cli-kit/src/session/authorize.ts#L20) in our codebase, what if we randomize that port and eliminate this scenario altogether? It's a bit odd that an error tells me to look up the process that's using a port. How do I do that?

There is any kind of validation on the `redirect_uri` made by the `identity` project.
image

**Comment by pepicrft** [comment]
> Don't you think it is a bit risky to let the user finishing a process (that maybe it's not the CLI) with so little information about it?

I don't think so. We can include additional information in the prompt to help developers make an informed decision. I think that's a better experience than asking users to figure out the process themselves. If I got that question I'd go straight to Google and I'd probably land on a Stackoverflow issue with a Unix command to run on my environment. If that's a yes/no question from the workflow, we save developers a lot of time.

**Comment by isaacroldan** [comment]
Right now the CLI can only use port 3456 for auth because it is the only one included in the internal configuration, but maybe we could add `backup` ports there if we want and try with those before failing.

**Comment by alvaro-shopify** [comment]
@isaacroldan I guess that this scenario is not very frequent and as we will need a verification on the blocked port/s maybe it does not worth to ask for the rights to add other supported redirection ports
I'm go with @pepicrft suggestion about the prompt. I will use the text proposed for the implementation and then @MeredithCastile could review and change it.

**Comment by alvaro-shopify** [comment]
@richard-shopify @isaacroldan @alvaro-shopify Modified the original behaviour that aborts the authorization process if the port is blocked.
Now, in case the port is blocked the user will be prompted to terminate it

**Review by pepicrft — APPROVED** [review]

File: packages/cli-kit/src/ui.ts [hunk 1]
```diff
@@ -141,8 +141,29 @@ export async function nonEmptyDirectoryPrompt(directory: string) {
...
+  ]
+
+  const choice = await prompt([
+    {
+      type: 'select',
+      name: 'value',
+      message: `The ${
+        description ?? 'process'
+      } requires the port ${port} to be available but it is currently being used by another process.
+      Would you like us to terminate that process to proceed with the authentication? `,
```
[the remark this answers is not in the document]
Comment (alvaro-shopify) [reply packages/cli-kit/src/ui.ts:157]: In fact I have tried to make this method reusable....but it seems that I forgot the second part.

**Comment by alvaro-shopify** [comment]
> Authentication requires a port (23456) that's unavailable because it's running another process (NAME_OF_PROCESS_HERE). Terminate that process?
> 
> * Yes,  terminate that process in order to log in now
> * No, cancel `<name of CLI command that triggered the auth>` and try later

Thank you for your feedback @MeredithCastile . We have two problems about gathering the information to change the message to the one you have proposed:
* (NAME_OF_PROCESS_HERE) - It is difficult to obtain the name of the process that is currently listening to the port `3456`. It is very dependant on the operating system and until now I haven't found a library cross platform that offers this feature
* name of CLI command that triggered the auth - Authentication process is very nested within the executions of the commands. We will have to do a big refactor to carry the command type to the authentication process.
What do you think about leaving the copy with a more generic text right now and create an issue to achieve those two goals?

Based completely on your proposal the temporary message could be:
image

**Comment by isaacroldan** [comment]
I'd say it's impossible to know the process or the cli command, we should just make the copy a bit more generic.

File: packages/cli-kit/src/session/authorize.ts [hunk 2]
```diff
@@ -54,3 +58,15 @@ export async function authorize(scopes: string[], state: string = randomHex(30))
...
+async function validateRedirectionPortAvailability(port: number) {
+  const result = await checkPort(port)
+  if (result === false) {
+    const cancelProcess = await terminateBlockingPortProcessPrompt(port, 'authentication')
+    if (cancelProcess) {
+      await killPortProcess(port)
+    } else {
+      throw new CancelExecution()
+    }
+  }
```
Comment (isaacroldan) [inline packages/cli-kit/src/session/authorize.ts:71]: you can do an early return here and avoid indentation.
```ts
if (result) return
```

**Comment by alvaro-shopify** [comment]
> I'd say it's impossible to know the process or the cli command, we should just make the copy a bit more generic.

I have finally found a library [find-process](https://github.com/yibn2008/find-process) to inform the user about the process name that is blocking the port. I have updated the PR description with a new prompt screenshot

On the other hand, as @isaacroldan has commented, accessing at this point to the command executed could be out of scope of the issue. I have left that part of the message using a generic `command` word. 

Please @MeredithCastile , could you confirm the message proposed is valid for now?

File: packages/cli-kit/src/ui.ts [hunk 3]
```diff
@@ -182,8 +183,34 @@ export async function nonEmptyDirectoryPrompt(directory: string) {
   }
 }
 
+export async function terminateBlockingPortProcessPrompt(port: number, stepDescription?: string): Promise<boolean> {
+  const stepDescriptionContent = stepDescription ?? 'current step'
+
+  const processInfo = await findProcess('port', port)
```
[the remark this answers is not in the document]
Comment (alvaro-shopify) [reply packages/cli-kit/src/ui.ts:189]: @isaacroldan  I have tested in the three platforms and it seems to work properly. What do you think about merging the PR?

File: packages/cli-kit/src/ui.ts [hunk 4]
```diff
@@ -182,8 +183,34 @@ export async function nonEmptyDirectoryPrompt(directory: string) {
...
+  const options = [
+    {name: 'Yes, terminate process in order to log in now', value: 'finish'},
+    {name: `No, cancel command and try later`, value: 'cancel'},
+  ]
+
+  const choice = await prompt([
+    {
+      type: 'select',
+      name: 'value',
+      message: `${stepDescriptionContent} requires a port ${port} that's unavailable because it's runnig another process${formattedProcessName}. Terminate that process? `,
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/cli-kit/src/ui.ts:204]: This is still not fixed! don't forget about it :)

File: packages/cli-kit/src/session/authorize.test.ts [hunk 5]
```diff
@@ -57,4 +65,45 @@ describe('authorize', () => {
...
+      codeChallenge: 'challenge',
+      codeVerifier: 'verifier',
+    }
+    vi.mocked(checkPort).mockResolvedValue(port)
+    vi.mocked(randomHex).mockReturnValue('hex')
+    vi.mocked(generateRandomChallengePair).mockReturnValue(challenge)
+    vi.mocked(listenRedirect).mockResolvedValue({code: 'code', state: 'state'})
+    vi.mocked(identity).mockResolvedValue('fqdn.com')
+    vi.mocked(clientId).mockResolvedValue('clientId')
+    vi.mocked(checkPort).mockResolvedValue(false)
```
Comment (isaacroldan) [inline packages/cli-kit/src/session/authorize.test.ts:94]: you are mocking `checkPort` twice here, I guess this is not intended and only the bottom one will be applied?

**Review by isaacroldan — APPROVED** [review]
Some minor comments and ready to merge

**Pull request merged**